# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [1]:
# Your code here
import pandas as pd

In [2]:
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)

In [21]:
df

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...
4003,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4004,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4006,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
# DataFrame shape
df.shape


(4008, 11)

- The dataset contains 4008 rows and 11 columns

In [6]:
# Data types of the columns
df.dtypes
print(df.dtypes)

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object


- Numerical columns are mostly correctly typed as int64 or float64. Some categorical variables are stored as object. If any numeric data is stored as object, it should be converted using pd.to_numeric(). 

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [37]:
# Checking for missing values

print("\nMissing Values:\n", df.isnull().sum())


Missing Values:
 Customer                     2937
ST                           2937
GENDER                       3054
Education                    2937
Customer Lifetime Value      2940
Income                       2937
Monthly Premium Auto         2937
Number of Open Complaints    2937
Policy Type                  2937
Vehicle Class                2937
Total Claim Amount           2937
dtype: int64


In [39]:
df.isnull().sum()

Customer                     2937
ST                           2937
GENDER                       3054
Education                    2937
Customer Lifetime Value      2940
Income                       2937
Monthly Premium Auto         2937
Number of Open Complaints    2937
Policy Type                  2937
Vehicle Class                2937
Total Claim Amount           2937
dtype: int64

In [ ]:
# Checking for duplicated rows

print("\nDuplicated Rows:", df.duplicated().sum())


Duplicated Rows: 2936


In [38]:
df.duplicated().sum()

np.int64(2936)

In [ ]:
# Checking for the unique values per column

df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [ ]:
# Identifying categorical columns

categorical_cols = df.select_dtypes(include="object").columns
categorical_cols

Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Number of Open Complaints', 'Policy Type', 'Vehicle Class'],
      dtype='object')

In [ ]:
# Checking for categorical values

for col in categorical_cols:
    print(f"\n{col}:")
    print(df[col].value_counts())


Customer:
Customer
MY31220    1
RB50392    1
QZ44356    1
AI49188    1
WW63253    1
          ..
FL50705    1
WC83389    1
OE15005    1
FV94802    1
BW63560    1
Name: count, Length: 1071, dtype: int64

ST:
ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

GENDER:
GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64

Education:
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64

Customer Lifetime Value:
Customer Lifetime Value
251459.20%    4
445811.34%    4
578018.22%    3
684615.03%    3
477294.38%    3
             ..
251753.36%    1
532667.77%    1
260027.21%    1
853479.28%    1
882883.50%    1
Name: count, Length: 1027, dtype: int64

Number of Open Complaint

- Columns with a small number of unique values are treated as categorical. These include [list columns]. Their distributions show [insight: dominant categories, imbalance, etc.].

In [19]:
df.describe()

,Income,Monthly Premium Auto,Total Claim Amount
count,1071.000000,1071.000000,1071.000000
mean,39295.701214,193.234360,404.986909
std,30469.427060,1601.190369,293.027260
min,0.000000,61.000000,0.382107
25%,14072.000000,68.000000,202.157702
50%,36234.000000,83.000000,354.729129
75%,64631.000000,109.500000,532.800000
max,99960.000000,35354.000000,2893.239678


In [18]:
df.describe(include="object")

,Customer,ST,GENDER,Education,Customer Lifetime Value,Number of Open Complaints,Policy Type,Vehicle Class
count,1071,1071,954,1071,1068,1071,1071,1071
unique,1071,8,5,6,1027,6,3,6
top,MY31220,Oregon,F,Bachelor,251459.20%,1/0/00,Personal Auto,Four-Door Car
freq,1,320,457,324,4,830,780,576


In [20]:
for col in categorical_cols:
    print(f"\n{col}:")
    print(df[col].value_counts(normalize=True))


Customer:
Customer
MY31220    0.000934
RB50392    0.000934
QZ44356    0.000934
AI49188    0.000934
WW63253    0.000934
             ...   
FL50705    0.000934
WC83389    0.000934
OE15005    0.000934
FV94802    0.000934
BW63560    0.000934
Name: proportion, Length: 1071, dtype: float64

ST:
ST
Oregon        0.298786
California    0.197012
Arizona       0.173669
Cali          0.112045
Nevada        0.091503
Washington    0.075630
WA            0.028011
AZ            0.023343
Name: proportion, dtype: float64

GENDER:
GENDER
F         0.479036
M         0.432914
Male      0.040881
female    0.029350
Femal     0.017820
Name: proportion, dtype: float64

Education:
Education
Bachelor                0.302521
College                 0.292250
High School or Below    0.276377
Master                  0.087768
Doctor                  0.034547
Bachelors               0.006536
Name: proportion, dtype: float64

Customer Lifetime Value:
Customer Lifetime Value
251459.20%    0.003745
445811.34%    0.00

**Numerical Statistics**
- The mean and median values suggest [normal/skewed] distributions. High standard deviation in some variables indicates variability. Potential outliers are observed in columns where max values are significantly higher than the median.

**Categorical Statistics**

- The most frequent categories dominate the dataset, suggesting possible imbalance. This may affect modeling and should be handled carefully.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [41]:
df["ST"].value_counts()

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

In [ ]:
# Your code here
# Count frequencies of customer locations
location_counts = df["ST"].value_counts()


# Get 5 least common locations (ascending order)
least_common = location_counts.sort_values(ascending=True).head(5)

least_common

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [28]:
# Your code here
# Count number of policies per type
policy_counts = df["Policy Type"].value_counts()

# Get the policy type with the highest count
top_policy = policy_counts.idxmax()

print(policy_counts)
print("\nTop policy type:", top_policy)


Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

Top policy type: Personal Auto


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [49]:
# Your code here

# Create filtered DataFrames using loc

personal_auto = df.loc[df["Policy Type"] == "Personal Auto"]
corporate_auto = df.loc[df["Policy Type"] == "Corporate Auto"]

# Display the first few rows of each filtered DataFrame

print("Personal Auto Policies:")
print(personal_auto.head())
print("\nCorporate Auto Policies:")
print(corporate_auto.head())

Personal Auto Policies:
  Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
4  GA49547  Washington      M  High School or Below              536307.65%   
5  OC83172      Oregon      F              Bachelor              825629.78%   

    Income  Monthly Premium Auto Number of Open Complaints    Policy Type  \
0      0.0                1000.0                    1/0/00  Personal Auto   
1      0.0                  94.0                    1/0/00  Personal Auto   
2  48767.0                 108.0                    1/0/00  Personal Auto   
4  36357.0                  68.0                    1/0/00  Personal Auto   
5  62902.0                  69.0                    1/0/00  Personal Auto   

   Vehicle Class  Total Claim Amount  

In [47]:
# Calculate average income
personal_income_avg = round(personal_auto["Income"].mean(), 2)
corporate_income_avg = round(corporate_auto["Income"].mean(), 2)

# Print results
print("Average Income (Personal Auto):", personal_income_avg)
print("Average Income (Corporate Auto):", corporate_income_avg)

Average Income (Personal Auto): 38180.7
Average Income (Corporate Auto): 41390.31


In [46]:
# Alternative print format

print(f"Average Income (Personal Auto): {personal_income_avg:.2f}")
print(f"Average Income (Corporate Auto): {corporate_income_avg:.2f}")

Average Income (Personal Auto): 38180.70
Average Income (Corporate Auto): 41390.31


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code here
# Reviewing statistics

df["Total Claim Amount"].describe()

count    1071.000000
mean      404.986909
std       293.027260
min         0.382107
25%       202.157702
50%       354.729129
75%       532.800000
max      2893.239678
Name: Total Claim Amount, dtype: float64

In [52]:
# Checking for the 75th percentile of the Total Claim Amount

q75 = df["Total Claim Amount"].quantile(0.75)
print("75th Percentile of Total Claim Amount:", q75)

75th Percentile of Total Claim Amount: 532.8


In [53]:
q25 = df["Total Claim Amount"].quantile(0.25)
print("25th Percentile of Total Claim Amount:", q25)

25th Percentile of Total Claim Amount: 202.157702


- The 75 quartile is the threshold value which means the top 25% has value greater than the output of **532.8** 

In [54]:
# Filtering out claims above the 75th percentile

filtered_high_claim_df = df[df["Total Claim Amount"] > q75]   

In [55]:
# Summary statistics for the high claim group

filtered_high_claim_df["Total Claim Amount"].describe()

count     264.000000
mean      782.228263
std       292.751640
min       537.600000
25%       606.521741
50%       679.597985
75%       851.400000
max      2893.239678
Name: Total Claim Amount, dtype: float64

- Customers with a total claim amount above the 75th percentile represent the top 25% of claims. These customers have significantly higher claim values compared to the rest of the dataset. The summary statistics show that the mean and maximum claim amounts in this group are substantially higher, indicating a concentration of high-cost claims. This segment may require special attention for risk management and profitability optimization.